### PR TITLE
CLD-5783 Implement rolling upgrade tests

### PR DESCRIPTION
**Pull request by mvitale1989**
#### Summary

This PR implements rolling release tests.

~~Note that as of now, they cannot work against `v9.10.x` or older versions (including the 9.5 ESR) due to recent changes in the tests themselves (see [this ticket for more details](https://mattermost.atlassian.net/browse/CLD-7983)). But they can be safely tested against the 9.11 RC images (e.g. from `9.11.0-rc1` to `9.11.0-rc3`).~~ This is now supported, testing from the 9.5.8 release works.

Overview of changes:
- Add `ROLLINGRELEASE_FROM_TAG` input to `e2e-fulltests-ci.yml`: if set, it enables a Rolling Release upgrade test, starting from the tag specified in this parameter.
  * The `ROLLINGRELEASE_commit_sha` and `ROLLINGRELEASE_SERVER_IMAGE` inputs if `e2e-test-ci-template.yml` are the way to communicate to the E2E test template if you want to perform a rolling release test. Only the latter input is used: the RollingRelease commit SHA will be used in the future, to select a commit to read the E2E testcases from.
- Changes in how the cloud user is managed locally:
  * Cloud user creation and deletion is not implied in e.g. a simple `make` invocation: it has to be done explicitly
  * The lifecycle of the cloud user is separated from the local instance's lifecycle. This allows you to create a single user, and use it for e.g. rolling release upgrade tests.
  * We don't create a new user if it's already been created in the past, locally, unless it's eventually deleted with `make cloud-teardown`.
- Make `make prepare-server` idempotent, to optimize the RollingRelease test implementation.
- If calling `make start-server` multiple times, only restart the containers that change. This is essential for RollingRelease tests, since we only want to change the `server` and `cypress` containers across the initial smoketests, and later the full E2E test.
- Fix server-to-E2Etest versioning issue: the `cypress/` directory is checked out to match the codebase of the Mattermost release that we're upgrading from, during RollingRelease testing.
- Improve reporting: show Automation Dashboard URL and, if applicable, the tag we're performing a Rolling Release upgrade test from.

#### Ticket Link

https://mattermost.atlassian.net/browse/CLD-5783
https://mattermost.atlassian.net/browse/CLD-7983

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by mvitale1989** [comment]
:point_up: the above was triggered manually from the [e2e-fulltests-ci.yml GHA page](https://github.com/mattermost/mattermost/actions/workflows/e2e-fulltests-ci.yml).

It seems to be working well: you can see in the `ci/e2e-tests` step of the E2E test workers (e.g. [this job](https://github.com/mattermost/mattermost/actions/runs/10251879109/job/28361039268)) that first there was the smoketest phase on the `v9.11.0-rc1` image, without using the Automation Dashboard, and then the usual full E2E test was run, using the `f9e1964098ca48191e23ca4adfb409615613cdef` commit's image. The [Automation Dashboard cycle](https://automation-dashboard.vercel.app/cycle/10251879109) only reports results from the full E2E run.

I see that there are some failures in the full E2E run to take a look at, i will do so tomorrow

**Comment by saturninoabril** [comment]
> Note that as of now, they cannot work against v9.10.x or older versions (including the 9.5 ESR) due to recent changes in the tests themselves (see [this ticket for more details](https://mattermost.atlassian.net/browse/CLD-7983))

I agree about the test case versioning (glad you pointed it out!), and would be great starting 9.11 -- being an ESR which will become the basis of future rolling upgrade tests. No need to address most recent non-ESR releases (such as 9.6 to 9.10). While it would be beneficial to include 9.5, if it requires significant additional work or backporting, it's fine to leave it as is.

File: .github/workflows/e2e-tests-ci-template.yml [hunk 1]
```diff
@@ -55,6 +55,12 @@ on:
       REPORT_TYPE:
         type: string
         required: false
+      ROLLINGRELEASE_commit_sha: # Unused for now, can be useful for implementing https://mattermost.atlassian.net/browse/CLD-7983
```
Comment (saturninoabril) [inline .github/workflows/e2e-tests-ci-template.yml:58]: 👍

File: .github/workflows/e2e-tests-ci-template.yml [hunk 2]
```diff
@@ -287,7 +294,25 @@ jobs:
...
+            make
+            )
+            echo "RollingRelease: asserting smoketest result has zero failures."
+            FAILURES=$(find cypress/results -name '*.json' | xargs -l jq -r '.stats.failures' | jq -s add)
+            if [ "$FAILURES" -ne "0" ]; then
+              echo "RollingRelease: initial smoketest for rolling release E2E run has nonzero ($FAILURES) failures. Aborting test run." >&2
+              exit 1
+            fi
+            rm -rfv cypress/results
+          fi
```
Comment (saturninoabril) [inline .github/workflows/e2e-tests-ci-template.yml:313]: Love this, on how you made it easier to reuse base setup/tests!

Comment (saturninoabril) [reply .github/workflows/e2e-tests-ci-template.yml:313]: Q: Is this cloud only? If yes, we should have it available for on-prem as well.

File: .github/workflows/e2e-fulltests-ci.yml [hunk 3]
```diff
@@ -13,6 +13,10 @@ on:
         type: string
         description: If testing a PR, this must be specified as well.
         required: false
+      ROLLINGRELEASE_FROM_TAG:
```
Comment (saturninoabril) [inline .github/workflows/e2e-fulltests-ci.yml:16]: nit: Can we have this as `ROLLING_RELEASE_<...>` (same with others)?
```suggestion
      ROLLING_RELEASE_FROM_TAG:
```

**Review by saturninoabril — APPROVED** [review]
This is awesome, thanks @mvitale1989! Overall, looks good to me except for some clarifications. Non-blocking and approve 💯

**Review by phoinixgrr — COMMENTED** [review]
Mostly unfamiliar with e2e testing, but I trust this is something you can ruthlessly  handle . 🚀

Regarding string manipulation in bash, I've put some general comments to simplify the logic. 

Awesome! 🎖️

File: .github/workflows/e2e-tests-ci-template.yml (hunk 2 above)
Comment (mvitale1989) [reply .github/workflows/e2e-tests-ci-template.yml:313]: It supports both cloud and on-prem. The `make cloud-init` and `make cloud-teardown` lines are just no-op if `SERVER=onprem`. The test run in this PR was for the onprem server, you can see that the cloud-specific lines are skipped [in the job output](https://github.com/mattermost/mattermost/actions/runs/10251879109/job/28361039268#step:5:40).

File: e2e-tests/.ci/server.prepare.sh [hunk 4]
```diff
@@ -26,16 +26,31 @@ if [ "$TEST" = "cypress" ]; then
...
-  ${MME2E_DC_SERVER} exec -T -- server curl -L --silent https://github.com/mattermost/mattermost-plugin-demo/releases/download/v0.8.0/com.mattermost.demo-plugin-0.8.0.tar.gz | ${MME2E_DC_SERVER} exec -T -u "$MME2E_UID" -- cypress tee tests/fixtures/com.mattermost.demo-plugin-0.8.0.tar.gz >/dev/null
   ${MME2E_DC_SERVER} exec -T -u "$MME2E_UID" -- cypress tee tests/fixtures/keycloak.crt >/dev/null <../../server/build/docker/keycloak/keycloak.crt
+  for PLUGIN_URL in \
+    "https://github.com/mattermost/mattermost-plugin-gitlab/releases/download/v1.3.0/com.github.manland.mattermost-plugin-gitlab-1.3.0.tar.gz" \
+    "https://github.com/mattermost/mattermost-plugin-demo/releases/download/v0.9.0/com.mattermost.demo-plugin-0.9.0.tar.gz" \
+    "https://github.com/mattermost/mattermost-plugin-demo/releases/download/v0.8.0/com.mattermost.demo-plugin-0.8.0.tar.gz"
+  do
+    PLUGIN_NAME=$(echo -n "$PLUGIN_URL" | rev | cut -d/ -f1 | rev)
```
[the remark this answers is not in the document]
Comment (mvitale1989) [reply e2e-tests/.ci/server.prepare.sh:35]: True that, thanks :smile: which made me think of the even simpler bash prefix removal, i'll commit it separately:
```
PLUGIN_NAME=${PLUGIN_URL##*/}
```

File: .github/workflows/e2e-tests-ci-template.yml (hunk 2 above)
Comment (saturninoabril) [reply .github/workflows/e2e-tests-ci-template.yml:313]: Got it, thanks!

**Review by phoinixgrr — APPROVED** [review]

File: e2e-tests/.ci/server.prepare.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (phoinixgrr) [reply e2e-tests/.ci/server.prepare.sh]: Ruthless! 🚀

**Pull request merged**